File: solutions/33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # soln - LeetCode one-pass binary search
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = l + (r-l)//2
            if nums[mid] == target:
                return mid
            elif nums[mid] >= nums[l]:
                if nums[mid] > target and target >= nums[l]:
                    r = mid - 1
                else:
                    l = mid + 1
            else: #nums[mid] < nums[l]
                if nums[mid] < target and target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
            # print(l,r, mid)
        return -1
        

File: solutions/33. Search in Rotated Sorted Array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size()-1, mid;
        while (l <= r) {
            mid = l + (r-l)/2;
            if (nums[mid] == target) return mid;
            else if (nums[mid] >= nums[l]) {
                if (nums[mid] > target && target >= nums[l]) r = mid-1;
                else l = mid+1;
            }
            else { //nums[mid] < nums[l]
                if (nums[mid] < target && target < nums[l]) l = mid+1;
                else r = mid-1;
            }
        }
        return -1;
    }
};
